### DISCUSSION TITLE
Connection Terminated: NO_ERROR

### DISCUSSION THREAD

**Discussion by splotsh** [encode/httpx#2032] [Q&A]
Hi,

I'm sending a few thousand requests (HTTP 1 and 2) across multiple servers and every once in a while, I see the following error:

`<ConnectionTerminated error_code:ErrorCodes.NO_ERROR, last_stream_id:2147483647, additional_data:None>`

I'm using an Async client with asyncio.
I've seen some old conversations about this, but I'm still not understanding what causes this issue. Any ideas on how I can dig in further to understand what is going on, or where the problem might be?

Thanks

**Comment by lovelydinosaur** [comment]
Well, that's an HTTP/2 case (it's got a stream ID) where the server has closed the connection.
Curiously, the stream ID is implausibly large on that message.

Not much else you can do with that other than retry the request.
I'd be interested in reproducible examples of this, just because it helps give broader context.

**Reply by splotsh** [reply]
Ok, will try and see how I can reliably reproduce this and come back if I find a way.

The stream ID is in the millions and in total I had sent off no more than 1000 requests per client instance. Stream ID is an incremental number that is initiated to 0 when a new client is created?

**Reply by lovelydinosaur** [reply]
> Stream ID is an incremental number that is initiated to 0 when a new client is created?

Yeah, with even numbers for streams in one direction, odd numbers for the other. (I don't recall which way around.)

It'd be interesting for us to have better debugging onto being able to inspect the HTTP/2 frames in both directions, in order to get a better idea of the underlying mechanics, and also to have better visibility when debugging gnarly behaviours.

**Reply by splotsh** [reply]
I have not been able to reproduce this reliably - it tends to happen on similar servers, but not always. I will report back if I am able to find something.

How would you suggest one detects that the connection has closed and sets one up again using asyncio?
